// @zaghmout/agent-sdk/src/index.ts
/**
 * © 2026 Flying Whale — ALL RIGHTS RESERVED
 * Owner: zaghmout.btc | ERC-8004 #54
 * https://fwgate.to
 *
 * @zaghmout/agent-sdk
 * Execution Control Plane for AI Agents on Stacks
 *
 * Nothing is considered executed unless FW_ETL_v1.0 attests it.
 */

const DEFAULT_BASE = 'https://fwgate.to';

// ── Types ─────────────────────────────────────────────────────────────────────

export interface EPCConfig {
  /** Override base URL — default: https://fwgate.to */
  baseUrl?: string;
  /** Request timeout in ms — default: 15000 */
  timeout?: number;
}

export interface Intent {
  /** Human-readable action: "swap 100 STX → ALEX" */
  action: string;
  /** Optional structured params */
  params?: Record<string, unknown>;
}

export interface AgentContext {
  /** Stacks address: SP... */
  address: string;
}

export interface ExecutionResult {
  /** Real Stacks txid — 64 hex chars */
  tx_hash: string;
  /** Optional execution metadata */
  meta?: Record<string, unknown>;
}

export interface EPCDecision {
  execution_allowed: boolean;
  confidence:        number;
  reason?:           string;
  cert_id?:          string;
  proof_url?:        string;
  [key: string]: unknown;
}

export interface AttestRecord {
  attest_id:      string;
  standard:       string;
  intent:         string;
  tx_hash:        string;
  agent:          string;
  chain_verified: boolean;
  tx_status:      string;
  block_height:   number | null;
  timestamp:      number;
  nonce:          string;
  signature:      string;
  proof_url:      string;
  issuer:         string;
}

export interface GuardResult {
  status:   'executed' | 'aborted';
  decision: EPCDecision;
  attest?:  AttestRecord;
  result?:  ExecutionResult;
}

// ── SDK factory ───────────────────────────────────────────────────────────────

export function createEPC1(config: EPCConfig = {}) {
  const base    = (config.baseUrl ?? DEFAULT_BASE).replace(/\/$/, '');
  const timeout = config.timeout ?? 15_000;

  async function post<T>(path: string, body: unknown): Promise<T> {
    const res = await fetch(`${base}${path}`, {
      method:  'POST',
      headers: { 'Content-Type': 'application/json' },
      body:    JSON.stringify(body),
      signal:  AbortSignal.timeout(timeout),
    });
    if (!res.ok) {
      const err = await res.text().catch(() => res.statusText);
      throw new Error(`FW_ETL_v1.0 [${res.status}] ${path}: ${err}`);
    }
    return res.json() as Promise<T>;
  }

  async function get<T>(path: string): Promise<T> {
    const res = await fetch(`${base}${path}`, {
      signal: AbortSignal.timeout(timeout),
    });
    if (!res.ok) {
      const err = await res.text().catch(() => res.statusText);
      throw new Error(`FW_ETL_v1.0 [${res.status}] ${path}: ${err}`);
    }
    return res.json() as Promise<T>;
  }

  return {
    /**
     * Pre-execution decision check (free, no payment required).
     * Returns execution_allowed + confidence score.
     * Always call this before executing anything.
     *
     * For full certified decision with replay protection:
     * use POST /epc/v1/evaluate (1000 μSTX) directly.
     */
    async verify(intent: Intent): Promise<EPCDecision> {
      return post<EPCDecision>('/epc/v1/check', {
        intent:    intent.action,
        agent:     intent.params?.agent,
        timestamp: Date.now(),
      });
    },

    /**
     * Core primitive — wraps execution with mandatory verification + attestation.
     *
     * Flow: verify → (abort if denied) → execute → attest → return proof
     *
     * The execution fn MUST return a real tx_hash from the chain.
     * No tx_hash = no attestation = execution is unproven.
     */
    async guard(
      intent:  Intent,
      fn:      () => Promise<ExecutionResult>,
      context: AgentContext,
    ): Promise<GuardResult> {
      const decision = await this.verify(intent);

      if (!decision.execution_allowed) {
        return { status: 'aborted', decision };
      }

      const result = await fn();

      if (!result?.tx_hash || !/^[a-f0-9]{64}$/i.test(result.tx_hash)) {
        throw new Error(
          'guard(): fn() must return { tx_hash: "<64-char-hex>" } ' +
          '— execution without a real txid cannot be attested'
        );
      }

      const attest = await this.attest(intent, result.tx_hash, context);

      return { status: 'executed', decision, result, attest };
    },

    /**
     * Issue an execution truth attestation.
     * Links intent + tx_hash + agent → Ed25519-signed certificate.
     * Stored in Redis, survives restarts, publicly verifiable.
     */
    async attest(
      intent:  Intent,
      tx_hash: string,
      context: AgentContext,
    ): Promise<AttestRecord> {
      return post<AttestRecord>('/epc/v1/attest', {
        intent: intent.action,
        tx_hash,
        agent:  context.address,
      });
    },

    /**
     * Retrieve an attestation by ID.
     */
    async getAttestation(attestId: string): Promise<AttestRecord> {
      return get<AttestRecord>(`/epc/v1/attest/${attestId}`);
    },

    /**
     * Live chain re-verification of an attestation.
     * Queries Stacks API fresh — does not rely on cached state.
     */
    async verify_attest(attestId: string): Promise<unknown> {
      return get(`/epc/v1/verify/${attestId}`);
    },

    /**
     * Full attestation history for an agent address.
     */
    async history(agentAddress: string): Promise<{
      agent:   string;
      count:   number;
      records: AttestRecord[];
    }> {
      return get(`/epc/v1/history/${agentAddress}`);
    },

    /**
     * Ed25519 public key — use this to verify signatures independently.
     */
    async pubkey(): Promise<{ pubkey_hex: string; algorithm: string }> {
      return get('/epc/v1/pubkey');
    },
  };
}

// ── Default singleton (fwgate.to) ─────────────────────────────────────────────
export const epc1 = createEPC1();
