// @zaghmout/agent-sdk/dist/index.js
// src/index.ts
var DEFAULT_BASE = "https://fwgate.to";
function createEPC1(config = {}) {
  const base = (config.baseUrl ?? DEFAULT_BASE).replace(/\/$/, "");
  const timeout = config.timeout ?? 15e3;
  async function post(path, body) {
    const res = await fetch(`${base}${path}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(timeout)
    });
    if (!res.ok) {
      const err = await res.text().catch(() => res.statusText);
      throw new Error(`FW_ETL_v1.0 [${res.status}] ${path}: ${err}`);
    }
    return res.json();
  }
  async function get(path) {
    const res = await fetch(`${base}${path}`, {
      signal: AbortSignal.timeout(timeout)
    });
    if (!res.ok) {
      const err = await res.text().catch(() => res.statusText);
      throw new Error(`FW_ETL_v1.0 [${res.status}] ${path}: ${err}`);
    }
    return res.json();
  }
  return {
    /**
     * Pre-execution decision check (free, no payment required).
     * Returns execution_allowed + confidence score.
     * Always call this before executing anything.
     *
     * For full certified decision with replay protection:
     * use POST /epc/v1/evaluate (1000 μSTX) directly.
     */
    async verify(intent) {
      return post("/epc/v1/check", {
        intent: intent.action,
        agent: intent.params?.agent,
        timestamp: Date.now()
      });
    },
    /**
     * Core primitive — wraps execution with mandatory verification + attestation.
     *
     * Flow: verify → (abort if denied) → execute → attest → return proof
     *
     * The execution fn MUST return a real tx_hash from the chain.
     * No tx_hash = no attestation = execution is unproven.
     */
    async guard(intent, fn, context) {
      const decision = await this.verify(intent);
      if (!decision.execution_allowed) {
        return { status: "aborted", decision };
      }
      const result = await fn();
      if (!result?.tx_hash || !/^[a-f0-9]{64}$/i.test(result.tx_hash)) {
        throw new Error(
          'guard(): fn() must return { tx_hash: "<64-char-hex>" } \u2014 execution without a real txid cannot be attested'
        );
      }
      const attest = await this.attest(intent, result.tx_hash, context);
      return { status: "executed", decision, result, attest };
    },
    /**
     * Issue an execution truth attestation.
     * Links intent + tx_hash + agent → Ed25519-signed certificate.
     * Stored in Redis, survives restarts, publicly verifiable.
     */
    async attest(intent, tx_hash, context) {
      return post("/epc/v1/attest", {
        intent: intent.action,
        tx_hash,
        agent: context.address
      });
    },
    /**
     * Retrieve an attestation by ID.
     */
    async getAttestation(attestId) {
      return get(`/epc/v1/attest/${attestId}`);
    },
    /**
     * Live chain re-verification of an attestation.
     * Queries Stacks API fresh — does not rely on cached state.
     */
    async verify_attest(attestId) {
      return get(`/epc/v1/verify/${attestId}`);
    },
    /**
     * Full attestation history for an agent address.
     */
    async history(agentAddress) {
      return get(`/epc/v1/history/${agentAddress}`);
    },
    /**
     * Ed25519 public key — use this to verify signatures independently.
     */
    async pubkey() {
      return get("/epc/v1/pubkey");
    }
  };
}
var epc1 = createEPC1();
export {
  createEPC1,
  epc1
};
